File: packages/core/src/marketplace/clawhub-adapter.ts
/**
 * ClawHub adapter — OpenClaw community marketplace
 * REST API driven, read operations need no auth
 */

import * as fs from 'fs';
import * as path from 'path';
import * as os from 'os';
import https from 'https';
import { MarketplaceSource, FetchedSkill, FetchOptions } from './types';

const CLAWHUB_API_BASE = 'https://clawdhub.com/api/v1';

export class ClawHubAdapter implements MarketplaceSource {
  name = 'clawhub';

  canHandle(url: string): boolean {
    return (
      url.startsWith('https://clawdhub.com/') ||
      url.startsWith('https://github.com/openclaw/') ||
      /^[a-z0-9-]+$/i.test(url) // Simple slug like "my-skill"
    );
  }

  async fetch(url: string, opts?: FetchOptions): Promise<FetchedSkill> {
    const cacheDir = opts?.cacheDir || path.join(os.homedir(), '.cache', 'skill-scanner', 'clawhub');
    const ttlMs = (opts?.cacheTtlHours || 24) * 60 * 60 * 1000;

    const slug = this.extractSlug(url);
    const destDir = path.join(cacheDir, slug);
    const metaPath = path.join(destDir, '.skill-scanner-meta.json');

    // Check cache
    if (!opts?.force && fs.existsSync(metaPath)) {
      try {
        const meta = JSON.parse(fs.readFileSync(metaPath, 'utf-8'));
        const age = Date.now() - new Date(meta.fetchedAt).getTime();
        if (age < ttlMs) {
          return this.buildFetchedSkill(destDir, url, slug, true);
        }
      } catch {
        // Cache invalid
      }
    }

    // Fetch from ClawHub API
    fs.mkdirSync(destDir, { recursive: true });

    // 1. Get skill metadata
    const skillMeta = await this.apiGet(`/skills/${slug}`);
    const skillName = skillMeta.name || slug;
    const version = skillMeta.version || 'latest';

    // 2. Download skill files
    const fileUrl = `${CLAWHUB_API_BASE}/skills/${slug}/file`;
    await this.downloadFile(fileUrl, destDir);

    // 3. Optionally fetch ClawHub's own scan results for cross-reference
    try {
      const clawhubScan = await this.apiGet(`/skills/${slug}/scan`);
      fs.writeFileSync(
        path.join(destDir, '.clawhub-scan.json'),
        JSON.stringify(clawhubScan, null, 2)
      );
    } catch {
      // ClawHub scan endpoint may not exist for all skills
    }

    // Write metadata
    fs.writeFileSync(
      metaPath,
      JSON.stringify(
        { fetchedAt: new Date().toISOString(), url, slug, version, skillName },
        null,
        2
      )
    );

    return this.buildFetchedSkill(destDir, url, skillName, false);
  }

  private extractSlug(url: string): string {
    if (url.startsWith('https://clawdhub.com/skills/')) {
      const match = url.match(/\/skills\/([^/?]+)/);
      return match ? match[1] : url;
    }
    if (url.startsWith('https://github.com/openclaw/')) {
      return url.split('/').pop() || url;
    }
    // Assume it's already a slug
    return url;
  }

  private async apiGet(endpoint: string): Promise<any> {
    return new Promise((resolve, reject) => {
      const url = `${CLAWHUB_API_BASE}${endpoint}`;
      https
        .get(url, { headers: { 'User-Agent': 'skill-scanner/0.1.0' } }, (res) => {
          let data = '';
          res.on('data', (chunk) => (data += chunk));
          res.on('end', () => {
            try {
              resolve(JSON.parse(data));
            } catch {
              reject(new Error(`Invalid JSON from ${url}`));
            }
          });
        })
        .on('error', reject)
        .setTimeout(15000, () => reject(new Error(`Timeout: ${url}`)));
    });
  }

  private async downloadFile(url: string, destDir: string): Promise<void> {
    return new Promise((resolve, reject) => {
      https
        .get(url, { headers: { 'User-Agent': 'skill-scanner/0.1.0' } }, (res) => {
          if (res.statusCode === 302 || res.statusCode === 301) {
            // Follow redirect
            const redirectUrl = res.headers.location;
            if (redirectUrl) {
              this.downloadFile(redirectUrl, destDir).then(resolve).catch(reject);
              return;
            }
          }

          // For v0.1, assume the file is a tarball or zip
          // In production, we'd stream to a temp file and extract
          let data = '';
          res.on('data', (chunk) => (data += chunk));
          res.on('end', () => {
            try {
              const files = JSON.parse(data);
              // Write files to destDir
              for (const [filePath, content] of Object.entries(files)) {
                const fullPath = path.join(destDir, filePath);
                fs.mkdirSync(path.dirname(fullPath), { recursive: true });
                fs.writeFileSync(fullPath, content as string);
              }
              resolve();
            } catch {
              // If not JSON, write as single file
              fs.writeFileSync(path.join(destDir, 'SKILL.md'), data);
              resolve();
            }
          });
        })
        .on('error', reject)
        .setTimeout(30000, () => reject(new Error(`Timeout: ${url}`)));
    });
  }

  private buildFetchedSkill(
    destDir: string,
    url: string,
    skillName: string,
    fromCache: boolean
  ): FetchedSkill {
    return {
      path: destDir,
      source: this.name,
      url,
      skillName,
      fetchedAt: new Date().toISOString(),
      fromCache,
      cleanup: async () => {
        // Keep cache by default
      },
    };
  }
}

File: crates/core/src/marketplace/clawhub_adapter.rs
use std::fs;
use std::path::Path;

use regex::Regex;

use super::types::{FetchOptions, FetchedSkill, MarketplaceSource};

const CLAWHUB_API_BASE: &str = "https://clawdhub.com/api/v1";
const MAX_REDIRECTS: u32 = 5;

pub struct ClawHubAdapter;

impl MarketplaceSource for ClawHubAdapter {
    fn name(&self) -> &str {
        "clawhub"
    }

    fn can_handle(&self, url: &str) -> bool {
        url.starts_with("https://clawdhub.com/")
            || url.starts_with("https://clawdhub.ai/")
            || url.starts_with("https://github.com/openclaw/")
            || Regex::new(r"^[a-zA-Z0-9-]+$").unwrap().is_match(url)
    }

    fn fetch(&self, url: &str, opts: &FetchOptions) -> Result<FetchedSkill, String> {
        let home = std::env::var("HOME").unwrap_or_else(|_| "/tmp".to_string());
        let cache_dir = opts
            .cache_dir
            .clone()
            .unwrap_or_else(|| format!("{}/.cache/skill-scanner/clawhub", home));
        let ttl_ms = opts.cache_ttl_hours.unwrap_or(24) * 60 * 60 * 1000;

        let slug = extract_slug(url);
        let dest_dir = format!("{}/{}", cache_dir, slug);
        let meta_path = format!("{}/.skill-scanner-meta.json", dest_dir);

        // Check cache
        if !opts.force && Path::new(&meta_path).exists() {
            if let Ok(content) = fs::read_to_string(&meta_path) {
                if let Ok(meta) = serde_json::from_str::<serde_json::Value>(&content) {
                    if let Some(fetched_at) = meta.get("fetchedAt").and_then(|v| v.as_str()) {
                        if let Ok(fetched_time) = chrono::DateTime::parse_from_rfc3339(fetched_at) {
                            let age =
                                chrono::Utc::now().signed_duration_since(fetched_time).num_milliseconds();
                            if (age as u64) < ttl_ms {
                                let skill_name = meta
                                    .get("skillName")
                                    .and_then(|v| v.as_str())
                                    .unwrap_or(&slug)
                                    .to_string();
                                return Ok(FetchedSkill {
                                    path: dest_dir,
                                    source: "clawhub".to_string(),
                                    url: url.to_string(),
                                    skill_name,
                                    fetched_at: chrono::Utc::now().to_rfc3339(),
                                    from_cache: true,
                                });
                            }
                        }
                    }
                }
            }
        }

        // Fetch from ClawHub API
        fs::create_dir_all(&dest_dir)
            .map_err(|e| format!("Failed to create directory: {}", e))?;

        // 1. Get skill metadata
        let meta_url = format!("{}/skills/{}", CLAWHUB_API_BASE, slug);
        let skill_meta_str = http_get(&meta_url, 0)?;
        let skill_meta: serde_json::Value = serde_json::from_str(&skill_meta_str)
            .map_err(|_| format!("Invalid JSON from {}", meta_url))?;
        let skill_name = skill_meta
            .get("name")
            .and_then(|v| v.as_str())
            .unwrap_or(&slug)
            .to_string();
        let version = skill_meta
            .get("version")
            .and_then(|v| v.as_str())
            .unwrap_or("latest")
            .to_string();

        // 2. Download skill files
        let file_url = format!("{}/skills/{}/file", CLAWHUB_API_BASE, slug);
        let file_data = http_get(&file_url, 0)?;
        if let Ok(files) = serde_json::from_str::<serde_json::Value>(&file_data) {
            if let Some(obj) = files.as_object() {
                for (file_path, content) in obj {
                    let full_path = format!("{}/{}", dest_dir, file_path);
                    if let Some(parent) = Path::new(&full_path).parent() {
                        let _ = fs::create_dir_all(parent);
                    }
                    if let Some(content_str) = content.as_str() {
                        let _ = fs::write(&full_path, content_str);
                    }
                }
            }
        } else {
            // If not JSON, write as single file
            let _ = fs::write(format!("{}/SKILL.md", dest_dir), &file_data);
        }

        // Write metadata
        let meta = serde_json::json!({
            "fetchedAt": chrono::Utc::now().to_rfc3339(),
            "url": url,
            "slug": slug,
            "version": version,
            "skillName": skill_name,
        });
        let _ = fs::write(&meta_path, serde_json::to_string_pretty(&meta).unwrap());

        Ok(FetchedSkill {
            path: dest_dir,
            source: "clawhub".to_string(),
            url: url.to_string(),
            skill_name,
            fetched_at: chrono::Utc::now().to_rfc3339(),
            from_cache: false,
        })
    }
}

fn extract_slug(url: &str) -> String {
    if url.starts_with("https://clawdhub.com/skills/") || url.starts_with("https://clawdhub.ai/skills/") {
        let re = Regex::new(r"/skills/([^/?]+)").unwrap();
        if let Some(cap) = re.captures(url) {
            return cap[1].to_string();
        }
    }
    if url.starts_with("https://github.com/openclaw/") {
        return url.split('/').next_back().unwrap_or(url).to_string();
    }
    // Assume it's already a slug
    url.to_string()
}

/// HTTP GET with redirect following (up to MAX_REDIRECTS)
fn http_get(url: &str, redirect_count: u32) -> Result<String, String> {
    if redirect_count > MAX_REDIRECTS {
        return Err(format!("Too many redirects (> {}) for {}", MAX_REDIRECTS, url));
    }

    // Use a simple blocking HTTP client via std::net
    // For production, reqwest would be used, but for the core library
    // we keep it simple with a basic implementation
    let _parsed = url::Url::parse(url).map_err(|e| format!("Invalid URL: {}", e))?;

    // For HTTPS, we need TLS. Since we don't have reqwest as a dep in core,
    // we'll use a simple approach that works for the adapter pattern.
    // In practice, the CLI crate would provide the HTTP implementation.
    // For now, return an error indicating network fetch is not available in core.
    Err(format!(
        "HTTP fetch not available in core library. Use CLI with reqwest for network operations. URL: {}",
        url
    ))
}
